src/storage/kv/store.rs

```diff
@@ -68,22 +84,82 @@ impl Store {
     /// Executes a function in a read-write transaction and commits the transaction.
     /// It begins a new read-write transaction, executes the function with the transaction, and commits the transaction.
     /// It returns the result of the function.
-    pub fn write(self: Arc<Self>, f: impl FnOnce(&mut Transaction) -> Result<()>) -> Result<()> {
-        let mut txn = self.begin_with_mode(Mode::ReadWrite)?;
+    pub async fn write(
+        self: Arc<Self>,
+        f: impl FnOnce(&mut Transaction) -> Result<()>,
+    ) -> Result<()> {
+        let mut txn = self.begin_with_mode(Mode::ReadWrite).await?;
         f(&mut txn)?;
-        txn.commit()?;
+        txn.commit().await?;
 
         Ok(())
     }
+
+    /// Closes the store. It sends a stop signal to the writer and waits for the done signal.
```

Comment:
Im not a storage guy, but does my thinking here work?

- storage must always be synchronous and safe via the WAL
- all WAL writes are synchronous and flushed
- all page writes can be asynchronous
- all page writes must be recoverable
- recovery of a single page is as simple as swapping pointers

So this close method is optional to avoid recovery cost. Do we think it's worth keeping? It is likely useful for tests etc where there are frequent starts/stops. But was wondering

Reply:
You are absolutely a storage guy with this thinking :)

So currently I am not committing to a WAL. The process is as follows:

1) Once a transaction is conflict free, it is selected for commit
2) The entries in this transaction are passed to a write channel as a `Task`
3) There is a thread which keeps receiving these tasks and commits them to first disk and then index
4) The Transaction is waiting on this task to get done to mark it as complete

This close method was just to ensure anything in the channel is not abruptly closed, whether I write to WAL or not.

Committing to WAL will help in case of restarts, to replay those transactions. It is for more durability guarantees. The WAL is already there, I'll add it as an optional step during precommit in a separate PR. The reason I haven't added it yet is because I wanted to take care of compaction on the WAL and the transaction log together. 

WDYT?

Reply:
I didnt realise we were choosing to not have a WAL. The WAL gives us a lot of benefits, even if it doesnt encompass a full system (i.e. just a shard).

- system backups can be incremental instead of always being full snapshot
- read replicas can lag behind leaders at any rate
- nodes can recover from a baseline snapshot and replay the WAL - this makes rejoining clusters easier
- streaming functionality heavily relies on WAL. Without it, there is no easy way to determine event ordering.

Reply:
Actually the transaction log is the ordered immutable log which can be used for most of the above. What I was talking about in terms of WAL is to add durability between precommit to commit. 

I agree, this will help in the replication/failover recovery, I am planning to add it post the async stuff